### ISSUE TITLE
needless_character_iteration false positive

### ISSUE THREAD

**Issue by aDotInTheVoid** [rust-lang/rust-clippy#12879]
[This transformation isn't correct:](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=b18e4300bb75c2cf248a272a4a490740)

```rust
fn main() {
    let s = "Beyoncé";
    // EVERY character isn't ascii
    dbg!(s.chars().all(|c| !char::is_ascii(&c)));
    // ANY character isn't ascii
    dbg!(!s.is_ascii());
}
```

```
[src/main.rs:3:5] s.chars().all(|c| !char::is_ascii(&c)) = false
[src/main.rs:4:5] !s.is_ascii() = true
```

_Originally posted by @aDotInTheVoid in https://github.com/rust-lang/rust-clippy/pull/12815#discussion_r1624071498_

CC @GuillaumeGomez

**Comment by GuillaumeGomez** [comment]
Thanks! Gonna send a fix in the next days.
